### ISSUE TITLE
choco uninstall all does not honor dependencies

### ISSUE THREAD

**Issue by mkevenaar** [chocolatey/choco#1822]
### What You Are Seeing?

`choco uninstall all -y` uninstalling all on alphabet 

### What is Expected?

Uninstalling honoring the dependencies

### How Did You Get This To Happen? (Steps to Reproduce)

**NOTE** Only try this on the chocolatey-test-environment or a test machine!

1. `choco install ci-cd-assets-vscode`
2. `choco uninstall all -y`

I would expect that it would uninstall
1. `ci-cd-assets-vscode`
2. `chocolatey-vscode.extension`
3. `vscode`

Actual order is 2,1,3 (including chocolatey, as expected)

**Comment by ferventcoder** [comment]
Does it error when it goes the wrong way?

**Comment by mkevenaar** [comment]
@ferventcoder as choco currently first uninstalls the extension, the POSH module and it’s commands are removed before the uninstallation of the vscode add-on package. 

In the add-on uninstaller, the one of the POSH module commandlets is used to remove the add-on from vscode.

**Comment by mkevenaar** [comment]
@ferventcoder I’ll try to make a gist soon.

**Comment by mkevenaar** [comment]
@ferventcoder https://gist.github.com/mkevenaar/9461720e59305b0b6602bda880fe2147

On line 554 you'll see the error message that `Uninstall-VsCodeExtension` is not found. It has been uninstalled on line 329.

/cc: @gep13

**Comment by ferventcoder** [comment]
Oh I think I understand. The packages required extensions in other packages that were removed first. So I don't think NuGet supports dependency order - you can force dependencies on uninstall if you want it to attempt proper order.

**Comment by mkevenaar** [comment]
@ferventcoder When you have things setup as mentioned above, and try to run `choco uninstall vscode` The installation tells you that you can't uninstall vscode as `ci-cd-assets-vscode` depends on it.

**Comment by ferventcoder** [comment]
> When you have things setup as mentioned above, and try to run choco uninstall vscode The installation tells you that you can't uninstall vscode as ci-cd-assets-vscode depends on it.

@mkevenaar that is correct behavior. You need to uninstall things in the proper order. Which is about what this issue is. What happens if you run `choco uninstall ci-cd-assets-vscode --force-dependencies`?

**Comment by mkevenaar** [comment]
@ferventcoder that was not the point I was trying to make here. 
I have given one example where I know it would go wrong when running `choco uninstall all`

What I think that has to be done, is calculate the dependencies first and then start uninstalling. 

I think this might be the location of the problem:
https://github.com/chocolatey/choco/blob/17f00eb9b98ef839d35a495e4b1a5ddabbc69c26/src/chocolatey/infrastructure.app/services/NugetService.cs#L1441-L1447

**Comment by ferventcoder** [comment]
Do you think removing the force dependencies would work? Something we could do without #508

**Comment by mkevenaar** [comment]
I think this needs to be , but it might work.

**Comment by VishalJadav** [comment]
Solution:

When you specify "all", chocolatey grabs packages from here:

https://github.com/chocolatey/choco/blob/17f00eb9b98ef839d35a495e4b1a5ddabbc69c26/src/chocolatey/infrastructure.app/services/NugetService.cs#L1686

https://github.com/chocolatey/choco/blob/17f00eb9b98ef839d35a495e4b1a5ddabbc69c26/src/chocolatey/infrastructure.app/services/NugetService.cs#L1690

The list "packagesToUpdate" needs to be ordered according to dependency flow.  This can be done via Topological sorting using depth-first search (DFS) algorithm.

**Comment by pauby** [comment]
I'm not able to reproduce this. As it's 3 and a bit years old now, I'm going to go ahead and close it. We can always reopen it again if necessary.
